miner/worker.go

```diff
@@ -676,31 +683,42 @@ func (w *worker) commitNewWork() {
 	txs := types.NewTransactionsByPriceAndNonce(w.current.signer, pending)
 	env.commitTransactions(w.mux, txs, w.chain, w.coinbase)
 
-	// Create the full block to seal with the consensus engine
-	fullState = env.state.Copy()
-	if fullBlock, err = w.engine.Finalize(w.chain, header, fullState, env.txs, uncles, env.receipts); err != nil {
-		log.Error("Failed to finalize block for sealing", "err", err)
-		return
-	}
+	w.finalize(uncles, w.fullTaskInterval, true)
+}
+
+// finalize runs any post-transaction state modifications, assembles the final block
+// and commits new work if consensus engine is running.
+func (w *worker) finalize(uncles []*types.Header, interval func(), update bool) error {
+	current := w.current
 	// Deep copy receipts here to avoid interaction between different tasks.
-	cpy := make([]*types.Receipt, len(env.receipts))
-	for i, l := range env.receipts {
-		cpy[i] = new(types.Receipt)
-		*cpy[i] = *l
+	receipts := make([]*types.Receipt, len(current.receipts))
+	for i, l := range current.receipts {
+		receipts[i] = new(types.Receipt)
+		*receipts[i] = *l
+	}
+
+	var (
+		block *types.Block
+		err   error
+	)
+	state := current.state.Copy()
+	if block, err = w.engine.Finalize(w.chain, current.header, state, current.txs, uncles, current.receipts); err != nil {
```

Comment:
I don't think there's value here to pre-declare `block` and `err`. You can just do

```go
block, err := w.engine.Finalize(w.chain, current.header, state, current.txs, uncles, current.receipts)
if err != nil {
	return err
}
```